datalad/support/sshconnector.py

```diff
@@ -450,7 +314,351 @@ def get_git_version(self):
+                'Failed to open SSH connection (could not start ControlMaster process)',
+                exit_code,
+                stdout,
+                stderr,
+            )
+        self._opened_by_us = True
+        return True
+
+    def close(self):
+        """Closes the connection.
+        """
+        if not self._opened_by_us:
+            lgr.debug("Not closing %s since was not opened by itself", self)
+            return
+        # stop controlmaster:
+        cmd = ["ssh", "-O", "stop"] + self._ssh_args + [self.sshri.as_str()]
+        lgr.debug("Closing %s by calling %s", self, cmd)
+        try:
+            self._runner.run(cmd, expect_stderr=True, expect_fail=True)
+        except CommandError as e:
+            lgr.debug("Failed to run close command")
+            if self.ctrl_path.exists():
+                lgr.debug("Removing existing control path %s", self.ctrl_path)
+                # socket need to go in any case
+                self.ctrl_path.unlink()
+            if e.code != 255:
+                # not a "normal" SSH error
+                raise e
+
+
+@auto_repr
+class BaseSSHManager(object):
+    """Interface for an SSHManager
+    """
+    def ensure_initialized(self):
+        """Assures that manager is initialized"""
+        pass
+
+    assure_initialized = ensure_initialized
+
+    def get_connection(self, url, use_remote_annex_bundle=True, force_ip=False):
+        """Get an SSH connection handler
+
+        Parameters
+        ----------
+        url: str
+          ssh url
+        force_ip : {False, 4, 6}
+          Force the use of IPv4 or IPv6 addresses.
+
+        Returns
+        -------
+        BaseSSHConnection
+        """
+        raise NotImplementedError
+
+    def _prep_connection_args(self, url):
+        # parse url:
+        from datalad.support.network import RI, is_ssh
+        if isinstance(url, RI):
+            sshri = url
+        else:
+            if ':' not in url and '/' not in url:
+                # it is just a hostname
+                lgr.debug("Assuming %r is just a hostname for ssh connection",
+                          url)
+                url += ':'
+            sshri = RI(url)
+
+        if not is_ssh(sshri):
+            raise ValueError("Unsupported SSH URL: '{0}', use "
+                             "ssh://host/path or host:path syntax".format(url))
+
+        from datalad import cfg
+        identity_file = cfg.get("datalad.ssh.identityfile")
+        return sshri, identity_file
+
+    def close(self, allow_fail=True):
+        """Closes all connections, known to this instance.
+
+        Parameters
+        ----------
+        allow_fail: bool, optional
+          If True, swallow exceptions which might be thrown during
+          connection.close, and just log them at DEBUG level
+        """
+        pass
+
+
+@auto_repr
+class NoMultiplexSSHManager(BaseSSHManager):
+    """
+    """
+
+    def __init__(self):
+        super().__init__()
```

Comment:
what is the point of having it? `@auto_repr` doesn't do anything to `__init__` and no `@borrowdoc` is used

Reply:
I just copied it. I don't know what it does.